linux/bindings/src/crypto/skcipher.rs

```diff
@@ -0,0 +1,264 @@
+    error::to_result,
+    new_mutex,
+    prelude::*,
+    str::CStr,
+    sync::Mutex,
+};
+
+use super::SgTable;
+
+/// A Symmetric Key Cipher wrapper.
+///
+/// The `crypto_skcipher` pointer is constructed by the `new` initializer.
+/// The recommended way to create such instances is with the [`pin_init`].
+///
+/// # Examples
+///
+/// The following is examples of creating [`Skcipher`] instances.
+///
+/// ```rust
+/// use kernel::{c_str, stack_try_pin_init};
+/// # use core::pin::Pin;
+/// # use crate::crypto::Skcipher;
+///
+/// // Allocates an instance on stack.
+/// stack_try_pin_init!(let foo = Skcipher::new(c_str!("cbc(aes)"), 0, 0));
+/// let foo: Result<Pin<&mut Skcipher>> = foo;
+///
+/// // Allocate an instance by Box::pin_init.
+/// let bar: Result<Pin<Box<Skcipher>>> = Box::pin_init(Skcipher::new(c_str!("cbc(aes)"), 0, 0));
+/// ```
+#[pin_data(PinnedDrop)]
+pub struct Skcipher {
+    #[pin]
+    write_access: Mutex<()>,
+    skcipher: NonNull<crate::crypto_skcipher>,
+}
+
+// SAFETY: `skcipher` is pointer to `crypto_skcipher`, which could be safely
+// transferred across threads.
+unsafe impl Send for Skcipher {}
+
+// SAFETY: `write_access` ensure the synchronization, so it could be
+// used concurrently from multiple threads.
+unsafe impl Sync for Skcipher {}
+
+impl Skcipher {
+    /// Constructs a new initializer.
+    ///
+    /// Try to allocate an skcipher handler, `alg_name` specifies the driver
+    /// name of the cipher, `type_` specifies the type of the cipher, `mask`
+    /// specifies the mask for the cipher.
+    ///
+    /// Returns a type impl [`PinInit<Skcipher>`] in case of success, or an error
+    /// code in [`Error`].
+    pub fn new(alg_name: &'static CStr, type_: u32, mask: u32) -> impl PinInit<Self, Error> {
+        try_pin_init!(Self {
+            write_access <- new_mutex!(()),
+            // SAFETY: `alg_name` has static lifetimes and live indefinitely.
+            // Any error will be checked by `IS_ERR`.
+            skcipher: unsafe {
+                let skcipher = crate::crypto_alloc_skcipher(alg_name.as_char_ptr(), type_, mask);
+                let const_ptr: *const core::ffi::c_void = skcipher.cast();
+                if IS_ERR(const_ptr) {
+                    let err = PTR_ERR(const_ptr) as i32;
+                    return Err(to_result(err).err().unwrap());
+                }
+                NonNull::new_unchecked(skcipher)
+            },
+        })
+    }
+
+    /// Allocates a cipher request, which could be used to handle the cipher
+    /// operations later.
+    pub fn alloc_request(&self) -> Result<SkcipherReq> {
+        // SAFETY: `self.skcipher` is a valid cipher.
+        let req = unsafe { crate::skcipher_request_alloc(self.skcipher.as_ptr(), GFP_KERNEL) };
+        if req.is_null() {
+            return Err(ENOMEM);
+        }
+
+        Ok(SkcipherReq {
+            // SAFETY: `req` is checked above, it is valid.
+            req: unsafe { NonNull::new_unchecked(req) },
```

Comment:
If `req` is a valid pointer, is `NonNull::new(req).expect("skcipher req ptr is null!");` a better choice? (Avoid an unsafe block and safety comment)

Reply:
That's a good point.